In [8]:
import sys
sys.path.append('../')
from  mongo.lib.mongo_db  import db_conn
import pandas as pd
import numpy as np
from pyecharts import Bar,configure,Pie
import time
from es_select import ES
import itertools


In [9]:
def date_to_time(date):
    timeArray = time.strptime(date, "%Y%m%d")
    time_semp = time.mktime(timeArray)
    return time_semp   # 2013
def date_to_date(date):
    timeArray = time.strptime(date, "%a %b %d %H:%M:%S %Y")
    time_semp = time.strftime("%Y%m%d%H%M%S", timeArray);
    return time_semp   # 2013

def time_to_today(timeStamp):
    timeArray = time.localtime(int(timeStamp))
    timeArray =time.strftime("%Y-%m-%d %H:%M:%S",timeArray)
    return timeArray   # 2013

## 1、查询20181106日联通出现 5017 （尊敬的客户您好，系统检测您的访问过于频繁，为保障您的账户安全请明天再试。）

In [10]:
data=db_conn.sid_info.find({"end_time":{'$gte':date_to_time('20181106'),'$lte':date_to_time('20181107')},'status':5017})
data1=list(data)

In [11]:
data1_rpt_pd=pd.DataFrame(x for x in data1)

## 2、将代理信息和目标URL取出，统计请求次数、使用代理数量、代理去重列表

In [12]:
def get_proxy(data1_rpt_pd):
    all_sid_http_proxy=[]
    for sid in data1_rpt_pd.sid.values.tolist():
        es=ES('20181107')
        one_sid_http_proxy={}
        #print (sid)
        ret_list=es.get_all_ret(sid)
        #print(len(ret_list))
        one_sid_http_proxy['sid']=sid
        one_sid_http_proxy['tel']=data1_rpt_pd.loc[data1_rpt_pd.sid.values==sid,'tel'].values.tolist()[0]
        one_sid_http_proxy['status']=data1_rpt_pd.loc[data1_rpt_pd.sid.values==sid,'status'].values.tolist()[0]
        one_sid_http_proxy['http_proxy']=[]
        one_sid_http_proxy['url_base']=[]
        one_sid_http_proxy['dates']=[]
        for i,one_http in enumerate(ret_list):
            one_sid_http_proxy['http_proxy'].append(one_http['_source']['MSG']['proxies']['http'].replace('http://','').split('.')[0])
            one_sid_http_proxy['url_base'].append(one_http['_source']['MSG']['url'])
            one_sid_http_proxy['dates'].append(date_to_date(one_http['_source']['DATETIME']))
        l = [(k, len(list(g))) for k, g in itertools.groupby(one_sid_http_proxy['http_proxy'])]
        one_sid_http_proxy['http_req_count']=l
        one_sid_http_proxy['req_count']=len(ret_list)
        one_sid_http_proxy['http_unique']=list(set(one_sid_http_proxy['http_proxy']))
        one_sid_http_proxy['http_unique_count']=len(one_sid_http_proxy['http_unique'])
        all_sid_http_proxy.append(one_sid_http_proxy)
    return all_sid_http_proxy
        

In [13]:
all_sid_http_proxy=get_proxy(data1_rpt_pd)

In [14]:
all_sid_http_proxy_pd=pd.DataFrame(x for x in all_sid_http_proxy) 

In [15]:
all_sid_http_proxy_pd=all_sid_http_proxy_pd.fillna(0)

## 3、将数据按照手机号码组织，并获取单个客户的请求此次，使用的代理数量

In [16]:
def chain_l(l):
    ret_dict={}
    for one in l:
        if one[0] in ret_dict:
            ret_dict[one[0]]+=one[1]
        else:
            ret_dict[one[0]]=one[1]
    return ret_dict

In [17]:
import itertools
def get_new_tel_list(all_sid_http_proxy_pd):
    new_tel_list=[]
    for tel in all_sid_http_proxy_pd.tel.unique().tolist():
        new_tel_dict={}
        tmp=all_sid_http_proxy_pd.loc[all_sid_http_proxy_pd.tel.values==tel,['tel','req_count','http_req_count','http_unique_count','http_unique']]
        tmp=tmp.fillna(0)
        #print(tmp.shape)
        new_tel_dict['tel']=tel
        new_tel_dict['req_count']=tmp.req_count.values.sum()
        #print(tmp.http_req_count.values.tolist())
        l=list(itertools.chain(*tmp.http_req_count.values.tolist()))
        l_dict=chain_l(l)
        new_tel_dict['http_req_count']=tuple(l_dict.values())
        new_tel_dict['http_unique_count']=tmp.http_unique_count.values.sum()
        new_tel_dict['http_unique']= tuple(l_dict)
        new_tel_list.append(new_tel_dict)
    new_tel_pd=pd.DataFrame(x for x in new_tel_list)
    return new_tel_pd

In [18]:
new_tel_pd=get_new_tel_list(all_sid_http_proxy_pd)

## 4、当天只使用同一个代理但是被封的客户

In [19]:
filte=np.logical_and(new_tel_pd.http_unique_count.values==1,True)
tmp_1=new_tel_pd.loc[filte,['tel','req_count','http_unique_count','http_unique']]

In [20]:
tmp_1

tel  req_count  http_unique_count    http_unique
2   18665287903        329                  1  (squidsz108,)
5   18685282281        134                  1   (squidsz75,)
6   15505996669        173                  1   (squidsz35,)
19  15172525372         13                  1   (squidsz34,)
21  18685432568        128                  1   (squidsz97,)
24  18272115317         13                  1   (squidsz94,)
25  15854686844         13                  1   (squidsz18,)
26  18685378880        125                  1   (squidsz13,)
29  15519044333        128                  1   (squidsz19,)
30  18671666468        124                  1   (squidsz84,)

### 发现  15172525372 ,15854686844 和 18272115317 存在异常值

In [21]:
data=db_conn.sid_info.find({"end_time":{'$gte':date_to_time('20181106'),'$lte':date_to_time('20181107')},'tel':{'$in':['15172525372','18272115317','15854686844']}})
data2=list(data)
data2=pd.DataFrame(x for x in data2)
one_sid_http_proxy=get_proxy(data2)
one_tel_http_proxy_pd=pd.DataFrame(x for x in one_sid_http_proxy) 
one_tel_http_proxy_pd=get_new_tel_list(one_tel_http_proxy_pd)

In [22]:
one_tel_http_proxy_pd.loc[:,['tel','req_count','http_unique_count','http_unique']]

tel  req_count  http_unique_count   http_unique
0  15172525372         13                  1  (squidsz34,)
1  15854686844         13                  1  (squidsz18,)
2  18272115317         13                  1  (squidsz94,)

### 以上三个请求均为移动

### 接下来分析 代理squidsz34，在这次请求前后是否正常

In [24]:
es=ES('20181107')
proxy_ret=es.get_proxy_ret('http://squidsz34.dianhua.cn:8080')
del es

In [25]:
one_sid_http_proxy_list=[]
for one in proxy_ret:
    one_sid_http_proxy={}
    one_sid_http_proxy['http_proxy']='http://squidsz34.dianhua.cn:8080'
    one_sid_http_proxy['url_base']=(one['_source']['MSG']['url'])
    one_sid_http_proxy['dates']=(date_to_date(one['_source']['DATETIME']))
    one_sid_http_proxy['sid']=((one['_source']['MSG']['sid']))
    one_sid_http_proxy_list.append(one_sid_http_proxy)


In [26]:
def sid_get_err_code(sid):
    data=db_conn.sid_info.find_one({"sid":sid,},{'_id':0,'status':1})
    return data.get('status',1)

In [27]:
one_sid_http_proxy_pd=pd.DataFrame(x for x in one_sid_http_proxy_list)
#one_sid_http_proxy_pd.groupby('sid')['http_proxy'].size().reset_index()
ip_proxy_sort_pd=one_sid_http_proxy_pd.groupby('sid')['dates'].max().reset_index().sort_values(by="dates")
ip_proxy_sort_pd['status']=ip_proxy_sort_pd.sid.map(sid_get_err_code)

In [28]:
ip_proxy_sort_pd

sid           dates  status
46  SIDecafae786bda4c7b899abfe39aeb796f  20181106001905       0
42  SIDd09fb89a221b41df92371b6a6edb7667  20181106061904       0
7   SID16e6e652939249bcba76d4b13ba1b735  20181106074014       0
17  SID44d43d958fd84187904dd54eeb0ed868  20181106090805       0
26  SID72a869b8b7b74f2bb24b8346dd78a5ff  20181106091217       0
32  SIDa8e84ad3be1744199dbb1b8cb4035df5  20181106095507       0
34  SIDb43a869505dc47f68b171d21f47fa1b4  20181106100309       0
21  SID57d164fe490045629344ac45a54c5f3b  20181106101921       0
2   SID09de2d93f8fe4d348a712ff7b76eb583  20181106104220       0
25  SID714c660bcd4045e99efaf32d1a9d92a5  20181106105327    5017
47  SIDeebf887c5ccb4cea9ed2341922a93302  20181106111408       0
12  SID2ef00a39ea9544dbbb8f59079719ec6f  20181106114442    4116
9   SID1eaf0286fc4642e9adfa3745fa9d2a88  20181106120325       0
13  SID31e00fec2b8c4e76a456ef46ca6c0472  20181106120925    3003
15  SID4026a6d3a07b4d09bad1a5cae6f84aee  20181106123312       0
8   SID1b0f3d5888c84202851a984543c73e0f  20181106130109       0
16  SID4471c4081a8942a3aae1d73483abc355  20181106130648    4999
28  SID873510c3830e4cc28f30342b869e8eee  20181106132741       0
24  SID667ac151e0f64bb488d6c72d346f4239  20181106134825       0
30  SID893f44b444904c50846e0bcf7fda2e64  20181106135328    3003
33  SIDb10771173b1e44f9a6e92a198c885932  20181106140822       0
11  SID2d51f52c696c4cf5ad18639a2cd31473  20181106142446       0
20  SID56b17bf08ecd415ea7b894dd39c76c67  20181106143933       0
41  SIDcf2cfbdc433c4f4b8ff07927d4e2a7b3  20181106144145       0
1   SID08da6b20c69845189e88f321a6207e1f  20181106145817       0
3   SID10644baed620457585f24027f39c2be3  20181106151638       0
19  SID47577f00453a4ac685b7f2df31694194  20181106152931    4999
4   SID10bd3ff9582d45fb95b7e78cb59b5251  20181106153522    5017
29  SID875c46d2655540b6847ba86e868f0fec  20181106154924       0
27  SID78e71ec72ce348bc83d9d15c01fe21c1  20181106155045    4999
36  SIDbfc4f7ac58854d23867e325253744a4e  20181106161026       0
44  SIDdf163386b71a48338d7572f9b2006dab  20181106162132    5017
0   SID04babcf7823040359fa92d449a394436  20181106162834    3003
10  SID27611a327d984aca8f5d7f551993ad07  20181106165117       0
18  SID4650ba5cdc87491781dbc2eba7096018  20181106165203       0
23  SID5ee77a79f8444a75a9a1b52771d2457b  20181106171808       0
45  SIDe4cd89eea8f14bf3af92ffd493306c48  20181106172431       0
37  SIDc7a6623c174147fc8b3060ee28b65ed8  20181106175047       0
14  SID38d69de0893343f2a773cb2f3c19e777  20181106180339       0
31  SID8b11abfadac6418ba43a92aae3555e6b  20181106182205       0
6   SID153022f0b1f143d582ef001f6c8f05af  20181106190553       0
39  SIDcd711a0ede9f4d469c75d7b47fb05662  20181106190857    5017
22  SID58da084ae827465e840060686c9b1e7f  20181106195441    4002
43  SIDd9bc85c85bba406a93b700b238a234f8  20181106205037       0
5   SID133a9cc9e3c14f92ba3b74ec255c996f  20181106205043       0
40  SIDcf1f7fb3c50848bd94b68e8aea318d0e  20181106214450    4010
35  SIDbb876d01add1455dac123de1e44c3ca2  20181106224256       0
38  SIDca9b6a746de64c7ca373ee0a34acf5af  20181106230203    3003

#### 出现了请求频繁，该代理之后的其他会话中还是可以正常使用，并爬取成功，怀疑可能该客户当天通过其他渠道操作过账户。

## 5、使用多个代理被封情况分析
req_count 为请求官网次数

http_req_count 为http_unique中代理对应请求次数

http_unique_count 为使用代理个数

http_unique 为被使用过的代理

### A、使用代理数量超过1个的客户

In [29]:
filte=np.logical_and(new_tel_pd.http_unique_count.values>1,True)
tmp_2=new_tel_pd.loc[filte,['tel','req_count','http_req_count','http_unique_count','http_unique']]
tmp_2

tel  req_count                           http_req_count  \
0   18649764213         15                       (3, 2, 2, 3, 4, 1)   
1   18580698151        190                                (128, 62)   
3   18685794878        151                                (15, 136)   
4   18507450873         17           (3, 3, 2, 3, 1, 1, 1, 1, 1, 1)   
7   15576882298         14                    (8, 1, 1, 1, 1, 1, 1)   
8   18138764589         20                             (5, 5, 5, 5)   
9   15963750395         26                                 (13, 13)   
10  13277091651          6                       (1, 1, 1, 1, 1, 1)   
11  13042237688        161                       (7, 2, 19, 132, 1)   
12  13192028199         32  (2, 7, 2, 2, 3, 2, 2, 2, 2, 2, 2, 2, 2)   
13  13096871866         10                             (3, 3, 3, 1)   
14  13118507914        140                          (2, 14, 2, 122)   
15  15585883123        141                                 (3, 138)   
16  18585210620        183                        (15, 15, 15, 138)   
17  18503653888         15           (3, 4, 1, 1, 1, 1, 1, 1, 1, 1)   
18  18609558955         14                    (3, 2, 3, 3, 1, 1, 1)   
20  13189238185          5                                (3, 1, 1)   
22  18853132281         24                                 (11, 13)   
23  18639521918         14                    (3, 3, 3, 2, 1, 1, 1)   
27  18553223366         13              (5, 1, 1, 1, 1, 1, 1, 1, 1)   
28  17584895319          6                       (1, 1, 1, 1, 1, 1)   
31  13103890371         17                 (3, 3, 2, 3, 3, 1, 1, 1)   
32  18253959596         59                     (11, 13, 11, 11, 13)   
33  13722666541          8                                   (4, 4)   
34  13294780204         11                          (5, 2, 2, 1, 1)   
35  13195877164         11                          (2, 2, 5, 1, 1)   
36  18666772645          7                                (5, 1, 1)   

    http_unique_count                                        http_unique  
0                   6  (squidsz47, squidsz44, squidsz62, squidsz43, s...  
1                   2                             (squidsz12, squidsz15)  
3                   2                           (squidsz124, squidsz122)  
4                  10  (squidsz25, squidsz30, squidsz28, squidsz49, s...  
7                   7  (squidsz04, squidsz30, squidsz13, squidsz14, s...  
8                   4       (squidsz68, squidsz65, squidsz74, squidsz11)  
9                   2                           (squidsz105, squidsz111)  
10                  6  (squidsz38, squidsz33, squidsz88, squidsz86, s...  
11                  5  (squidsz67, squidsz71, squidsz73, squidsz35, s...  
12                 13  (squidsz62, squidsz75, squidsz73, squidsz49, s...  
13                  4       (squidsz34, squidsz63, squidsz43, squidsz70)  
14                  4       (squidsz45, squidsz65, squidsz41, squidsz39)  
15                  2                            (squidsz130, squidsz16)  
16                  4      (squidsz97, squidsz71, squidsz06, squidsz128)  
17                 10  (squidsz70, squidsz68, squidsz19, squidsz20, s...  
18                  7  (squidsz09, squidsz12, squidsz15, squidsz18, s...  
20                  3                  (squidsz43, squidsz50, squidsz46)  
22                  2                           (squidsz117, squidsz126)  
23                  7  (squidsz99, squidsz106, squidsz112, squidsz102...  
27                  9  (squidsz11, squidsz25, squidsz37, squidsz26, s...  
28                  6  (squidsz35, squidsz34, squidsz47, squidsz49, s...  
31                  8  (squidsz95, squidsz97, squidsz111, squidsz114,...  
32                  5  (squidsz121, squidsz124, squidsz01, squidsz03,...  
33                  2                             (squidsz06, squidsz07)  
34                  5  (squidsz19, squidsz18, squidsz17, squidsz22, s...  
35                  5  (squidsz86, squidsz88, squidsz91, squidsz93, s... 

### B、分析代理总调用次数小于100次的客户

In [34]:
filte=np.logical_and(new_tel_pd.http_unique_count.values>1,new_tel_pd.req_count.values<100)
tmp_3=new_tel_pd.loc[filte,['tel','req_count','http_req_count','http_unique_count','http_unique']]
tmp_3

tel  req_count                           http_req_count  \
0   18649764213         15                       (3, 2, 2, 3, 4, 1)   
4   18507450873         17           (3, 3, 2, 3, 1, 1, 1, 1, 1, 1)   
7   15576882298         14                    (8, 1, 1, 1, 1, 1, 1)   
8   18138764589         20                             (5, 5, 5, 5)   
9   15963750395         26                                 (13, 13)   
10  13277091651          6                       (1, 1, 1, 1, 1, 1)   
12  13192028199         32  (2, 7, 2, 2, 3, 2, 2, 2, 2, 2, 2, 2, 2)   
13  13096871866         10                             (3, 3, 3, 1)   
17  18503653888         15           (3, 4, 1, 1, 1, 1, 1, 1, 1, 1)   
18  18609558955         14                    (3, 2, 3, 3, 1, 1, 1)   
20  13189238185          5                                (3, 1, 1)   
22  18853132281         24                                 (11, 13)   
23  18639521918         14                    (3, 3, 3, 2, 1, 1, 1)   
27  18553223366         13              (5, 1, 1, 1, 1, 1, 1, 1, 1)   
28  17584895319          6                       (1, 1, 1, 1, 1, 1)   
31  13103890371         17                 (3, 3, 2, 3, 3, 1, 1, 1)   
32  18253959596         59                     (11, 13, 11, 11, 13)   
33  13722666541          8                                   (4, 4)   
34  13294780204         11                          (5, 2, 2, 1, 1)   
35  13195877164         11                          (2, 2, 5, 1, 1)   
36  18666772645          7                                (5, 1, 1)   

    http_unique_count                                        http_unique  
0                   6  (squidsz47, squidsz44, squidsz62, squidsz43, s...  
4                  10  (squidsz25, squidsz30, squidsz28, squidsz49, s...  
7                   7  (squidsz04, squidsz30, squidsz13, squidsz14, s...  
8                   4       (squidsz68, squidsz65, squidsz74, squidsz11)  
9                   2                           (squidsz105, squidsz111)  
10                  6  (squidsz38, squidsz33, squidsz88, squidsz86, s...  
12                 13  (squidsz62, squidsz75, squidsz73, squidsz49, s...  
13                  4       (squidsz34, squidsz63, squidsz43, squidsz70)  
17                 10  (squidsz70, squidsz68, squidsz19, squidsz20, s...  
18                  7  (squidsz09, squidsz12, squidsz15, squidsz18, s...  
20                  3                  (squidsz43, squidsz50, squidsz46)  
22                  2                           (squidsz117, squidsz126)  
23                  7  (squidsz99, squidsz106, squidsz112, squidsz102...  
27                  9  (squidsz11, squidsz25, squidsz37, squidsz26, s...  
28                  6  (squidsz35, squidsz34, squidsz47, squidsz49, s...  
31                  8  (squidsz95, squidsz97, squidsz111, squidsz114,...  
32                  5  (squidsz121, squidsz124, squidsz01, squidsz03,...  
33                  2                             (squidsz06, squidsz07)  
34                  5  (squidsz19, squidsz18, squidsz17, squidsz22, s...  
35                  5  (squidsz86, squidsz88, squidsz91, squidsz93, s...  
36                  3                  (squidsz43, squidsz47, squidsz46)

#### 查看这些客户当日是否有其他返回码的请求

In [31]:
data=db_conn.sid_info.find({"end_time":{'$gte':date_to_time('20181106'),'$lte':date_to_time('20181107')},'tel':{'$in':tmp_3.tel.values.tolist()}})
data2=list(data)
data2=pd.DataFrame(x for x in data2)
some_sid_http_proxy=get_proxy(data2)
some_sid_http_proxy_pd=pd.DataFrame(x for x in some_sid_http_proxy) 
some_sid_http_proxy_pd=get_new_tel_list(some_sid_http_proxy_pd)

In [32]:
some_sid_http_proxy_pd

http_req_count  \
0                              (3, 3, 3, 1)   
1                  (3, 3, 2, 3, 3, 1, 1, 1)   
2                                 (3, 1, 1)   
3   (2, 7, 2, 2, 3, 2, 2, 2, 2, 2, 2, 2, 2)   
4                           (2, 2, 5, 1, 1)   
5                        (1, 1, 1, 1, 1, 1)   
6                           (5, 2, 2, 1, 1)   
7                                    (4, 4)   
8                     (8, 1, 1, 1, 1, 1, 1)   
9                                  (13, 13)   
10                       (1, 1, 1, 1, 1, 1)   
11                             (5, 5, 5, 5)   
12                     (11, 13, 11, 11, 13)   
13           (3, 4, 1, 1, 1, 1, 1, 1, 1, 1)   
14           (3, 2, 3, 3, 1, 1, 1, 1, 1, 1)   
15              (5, 1, 1, 1, 1, 1, 1, 1, 1)   
16                    (3, 2, 3, 3, 1, 1, 1)   
17                    (3, 3, 3, 2, 1, 1, 1)   
18                       (3, 2, 2, 3, 4, 1)   
19                                (5, 1, 1)   
20                                 (11, 13)   

                                          http_unique  http_unique_count  \
0        (squidsz34, squidsz63, squidsz43, squidsz70)                  4   
1   (squidsz95, squidsz97, squidsz111, squidsz114,...                  8   
2                   (squidsz43, squidsz50, squidsz46)                  3   
3   (squidsz62, squidsz75, squidsz73, squidsz49, s...                 13   
4   (squidsz86, squidsz88, squidsz91, squidsz93, s...                  5   
5   (squidsz38, squidsz33, squidsz88, squidsz86, s...                  6   
6   (squidsz19, squidsz18, squidsz17, squidsz22, s...                  5   
7                              (squidsz06, squidsz07)                  2   
8   (squidsz04, squidsz30, squidsz13, squidsz14, s...                  7   
9                            (squidsz105, squidsz111)                  2   
10  (squidsz35, squidsz34, squidsz47, squidsz49, s...                  6   
11       (squidsz68, squidsz65, squidsz74, squidsz11)                  4   
12  (squidsz121, squidsz124, squidsz01, squidsz03,...                  5   
13  (squidsz70, squidsz68, squidsz19, squidsz20, s...                 10   
14  (squidsz25, squidsz28, squidsz30, squidsz49, s...                 10   
15  (squidsz11, squidsz25, squidsz37, squidsz26, s...                  9   
16  (squidsz09, squidsz12, squidsz15, squidsz18, s...                  7   
17  (squidsz99, squidsz106, squidsz112, squidsz102...                  7   
18  (squidsz47, squidsz44, squidsz62, squidsz43, s...                  6   
19                  (squidsz43, squidsz46, squidsz47)                  3   
20                           (squidsz117, squidsz126)                  2   

    req_count          tel  
0          10  13096871866  
1          17  13103890371  
2           5  13189238185  
3          32  13192028199  
4          11  13195877164  
5           6  13277091651  
6          11  13294780204  
7           8  13722666541  
8          14  15576882298  
9          26  15963750395  
10          6  17584895319  
11         20  18138764589  
12         59  18253959596  
13         15  18503653888  
14         17  18507450873  
15         13  18553223366  
16         14  18609558955  
17         14  18639521918  
18         15  18649764213  
19          7  18666772645  
20         24  18853132281

#### 这些用户当日不存在其他请求，由此推测可能是频繁变更IP地址引起封禁

### C、分析请求数量超过100的代理

In [35]:
filte=np.logical_and(new_tel_pd.http_unique_count.values>1,new_tel_pd.req_count.values>100)
tmp_4=new_tel_pd.loc[filte,['tel','req_count','http_req_count','http_unique_count','http_unique']]
tmp_4

tel  req_count      http_req_count  http_unique_count  \
1   18580698151        190           (128, 62)                  2   
3   18685794878        151           (15, 136)                  2   
11  13042237688        161  (7, 2, 19, 132, 1)                  5   
14  13118507914        140     (2, 14, 2, 122)                  4   
15  15585883123        141            (3, 138)                  2   
16  18585210620        183   (15, 15, 15, 138)                  4   

                                          http_unique  
1                              (squidsz12, squidsz15)  
3                            (squidsz124, squidsz122)  
11  (squidsz67, squidsz71, squidsz73, squidsz35, s...  
14       (squidsz45, squidsz65, squidsz41, squidsz39)  
15                            (squidsz130, squidsz16)  
16      (squidsz97, squidsz71, squidsz06, squidsz128)

#### 使用代理数量相对B情况要少，但是均存在某一个代理使用超过100的情况，可能是因为当日请求量超过阈值或者Ip地址发送变化引起

### D、分析C中代理使用量超过100的情况发生的时机

In [36]:
def select_tel(tel):
    data=db_conn.sid_info.find({"end_time":{'$gte':date_to_time('20181106'),'$lte':date_to_time('20181107')},'tel':tel})
    data2=list(data)
    data2=pd.DataFrame(x for x in data2)
    some_sid_http_proxy=get_proxy(data2)
    some_sid_http_proxy_pd=pd.DataFrame(x for x in some_sid_http_proxy) 
    some_sid_http_proxy_pd=some_sid_http_proxy_pd.loc[some_sid_http_proxy_pd.req_count.values>0,['sid','req_count','tel','status','dates']]
    some_sid_http_proxy_pd['dates']=some_sid_http_proxy_pd.dates.map(max)
    return some_sid_http_proxy_pd

In [37]:
select_tel('18580698151')

sid  req_count          tel  status  \
0  SIDf7f785e67639491cb1bd8319cf33060a        128  18580698151    5017   
1  SID91066c1673d74ce6a2135da4fea438f7         62  18580698151    5017   

            dates  
0  20181106224212  
1  20181106224207

In [38]:
select_tel('13042237688')

sid  req_count          tel  status  \
0  SID22914fa2f81f4839ab44fc580044bc97        133  13042237688    5017   
1  SID4dfe4f17b47f497f98020164b919757c         28  13042237688    5017   

            dates  
0  20181106165850  
1  20181106170005

#### 由此可以看出，13042237688 先发生了133次官网请求，然后重试有发生了一个28次的官网请求，在首次请求时已经触发封禁

## 5、结论
### 1）、单次爬虫事务中，如果短时间内请求官网超过125次，可能被封禁
### 2）、多次爬虫事务中，如果使用多个ip代理请求官网，超过2个ip将可能被封禁
### 3）、联通反爬策略为，tel+ip进行识别，相同ip不同客户、相互不干扰
### 4）、针对分析结果，建立爬虫代理池对该以上反爬策略无效